### PR TITLE
feat: allow containerd two minor version upgrades

### PR DESCRIPTION
**Pull request by ricardomaraschini**
#### What this PR does / why we need it:

While Containerd does not support upgrades that span more than one minor release, we aim to enable users to upgrade between two minor Containerd releases. Therefore, this pull request introduces a constraint to the installation script that allows for such upgrades. Going forward, the installer will only permit Containerd upgrades that span two minor releases, and upgrades will be executed one at a time.

The table below displays some of the outputs resulting from this change.

| From | To   | outcome                                               |
|------|------|-------------------------------------------------------|
| v1.2 | v1.5 | The upgrade is not possible,  error message is shown. |
| v1.3 | v1.4 | Upgrade succeeds.                                     |
| v1.3 | v1.5 | Upgrades first to v1.4 and then to v1.5.                    |
| v1.4 | v1.3 | Downgrades are blocked, error message is shown.       |
| v1.6 | v2.0 | Major version upgrades are blocked                   |

##### About major version upgrades

As we can't understand how the migration to Containerd v2 will work we have blocked this kind of upgrade. The idea is to make `testgrid` to fail in such migrations (once Containerd v2 comes to live).

#### Special notes for your reviewer:

This is yet a draft. I need help in understanding how to generate the Manifest' `CONTAINERD_STEP_VERSIONS` variable automatically as well as how can I add `containerd_version - 1` to the installer.

### CODE REVIEW

File: hack/testdata/manifest/clean [hunk 1]
```diff
@@ -12,4 +12,6 @@ KURL_BIN_UTILS_FILE=
 STEP_VERSIONS=(0.0.0 0.0.0 0.0.0 0.0.0 0.0.0 0.0.0 0.0.0 0.0.0 0.0.0 0.0.0 0.0.0 0.0.0 0.0.0 0.0.0 0.0.0 0.0.0 1.16.4 1.17.13 1.18.20 1.19.16 1.20.15 1.21.14 1.22.17 1.23.17 1.24.12 1.25.8 1.26.3)
 # ROOK_STEP_VERSIONS array is generated by the server and injected at runtime based on supported rook versions
 ROOK_STEP_VERSIONS=(1.0.4-14.2.21 0.0.0 0.0.0 0.0.0 1.4.9 1.5.12 1.6.11 1.7.11 1.8.10 1.9.12 1.10.11 1.11.2)
+# CONTAINERD_STEP_VERSIONS array is generated by the server and injected at runtime based on supported containerd versions
+CONTAINERD_STEP_VERSIONS=(1.2.13 1.3.9 1.4.13 1.5.11 1.6.18)
```
Comment (camilamacedo86) [inline hack/testdata/manifest/clean:16]: You need to revert those changes

**Comment by ricardomaraschini** [comment]
> This is looking good.
> 
> You will want to make sure to automate the generation of the step versions

@emosbaugh Done.

File: testgrid/specs/full.yaml [hunk 2]
```diff
@@ -1744,3 +1744,79 @@
   unsupportedOSIDs:
   - ol-9 # containerd < 1.6 is not supported on rhel 9 variants
   - rocky-91 # containerd < 1.6 is not supported on rhel 9 variants
+- name: "containerd upgrade from 1.2.x to 1.4.x"
```
Comment (camilamacedo86) [inline testgrid/specs/full.yaml:1747]: have we not tests that we are updating more than 2 versions which will start to fail after we get this one merged?

Comment (ricardomaraschini) [reply testgrid/specs/full.yaml:1747]: If we have them they will fail and then we gonna address them.

File: scripts/common/containerd.sh [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+# is possible and in case it is, returns the list of steps necessary to perform the upgrade.
+# each step is a version of containerd that we need to install.
+export CONTAINERD_INSTALL_VERSIONS=()
+function containerd_evaluate_upgrade() {
+    local from_version=$1
+    local to_version=$2
+    echo "Evaluating if an upgrade from containerd v$from_version to v$to_version is possible."
+    containerd_upgrade_is_possible "$from_version" "$to_version"
+    echo "Containerd upgrade from v$from_version to v$to_version is possible."
+    for version in $(containerd_migration_steps "$from_version" "$to_version"); do
```
[the remark this answers is not in the document]
Comment (camilamacedo86) [reply scripts/common/containerd.sh:106]: Example:

from: https://testgrid.kurl.sh/run/pr-4304-d62a7c6-containerd-1.6.10-k8s-ctrd-2023-03-30T04:45:32Z
Screenshot 2023-03-30 at 13 55 15

File: scripts/common/common.sh [hunk 4]
```diff
@@ -424,7 +424,18 @@ function report_install_docker() {
 }
 
 function report_install_containerd() {
-    addon_install "containerd" "$CONTAINERD_VERSION"
+    if [ -z "$CURRENT_KUBERNETES_VERSION" ] || [ ! -f "/usr/bin/containerd" ]; then
```
Comment (rrpolanco) [inline scripts/common/common.sh:427]:
```suggestion
    if [ -z "$CURRENT_KUBERNETES_VERSION" ] || ! commandExists containerd; then
```
There's a built in command to check if command exist: `commandExists`.

**Review by rrpolanco — APPROVED** [review]
This looks right and I don't have any objections to this.

**Review by rrpolanco — APPROVED** [review]

**Review by rrpolanco — APPROVED** [review]

**Comment by ricardomaraschini** [comment]
Remarks on the the failing check in this PR: FOSSA scan failure is not because of a dependency issue in this PR but due to a backend problem on FOSSA server side: https://github.com/replicatedhq/kURL/actions/runs/4615501554/jobs/8159453877?pr=4291

```
>>> Relevant errors

    Error

      A connection to the FOSSA endpoint was established, but the service took too long to respond.

          Request: GET https://app.fossa.com/api/cli/custom+5995%2Fgithub.com%2Freplicatedhq%2Fkurl$d6cbb231a84e0e821ee083ea550dc04206f9ff3f/latest_build

      This is a networking error.

      Networking errors are typically caused by actual network failure or a network appliance
      (e.g. a firewall) between the FOSSA CLI and the FOSSA backend.
      This means that often such errors are transient, or are caused by local network configuration.

      Trying again in a few minutes may resolve this issue.
      If this issue persists, please report a bug to FOSSA support at https://support.fossa.com/

      In your bug report, please include FOSSA's debug bundle file: fossa.debug.json.gz.

      You can generate debug bundle by using `--debug` flag, for example:
        fossa analyze --debug

      Traceback:
        (none)
```

**Pull request merged**